### PR TITLE
ddl: modify different character sets varchar type column maximum length limit 

### PR DESCRIPTION
**Pull request by ciscoxll**
Cherry pick #8818

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8464)

### CODE REVIEW

**Comment by ciscoxll** [comment]
@winkyao @tiancaiamao  @crazycs520 PTAL

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -1902,6 +1947,13 @@ func (d *ddl) AlterTableCharsetAndCollate(ctx sessionctx.Context, ident ast.Iden
 		return errors.Trace(err)
 	}
 
+	for _, col := range tb.Meta().Cols() {
+		if col.Tp == mysql.TypeVarchar {
+			err = IsTooBigFieldLength(col.Flen, col.Name.O, toCharset)
+			return errors.Trace(err)
```
Comment (crazycs520) [inline ddl/ddl_api.go]: ???

**Comment by ciscoxll** [comment]
/run-all-tests tidb-test=release-2.1 tikv=release-2.1 pd=release-2.1

**Comment by ciscoxll** [comment]
/run-all-tests tidb-test=release-2.1 tikv=release-2.1 pd=release-2.1

**Pull request merged**